core/staking_verifier.go

```diff
@@ -0,0 +1,262 @@
+// does not use said components will still run.
+//
+// Note that these checks never update the stateDB, it only reads from stateDB.
+func VerifyCreateValidatorStakingMsg(
+	stateDB vm.StateDB, epoch *big.Int, msg *staking.CreateValidator,
+) error {
+	if msg.Amount.Sign() == -1 {
+		return errNegativeAmount
+	}
+	if stateDB != nil {
+		if stateDB.IsValidator(msg.ValidatorAddress) {
+			return errors.Wrapf(errValidatorExist, common2.MustAddressToBech32(msg.ValidatorAddress))
+		}
+		if !CanTransfer(stateDB, msg.ValidatorAddress, msg.Amount) {
+			return errInsufficientBalanceForStake
+		}
+	}
+	// create temp validator to do sanity check
+	v, err := staking.CreateValidatorFromNewMsg(msg, new(big.Int).SetInt64(0))
+	if err != nil {
+		return err
+	}
+	zero := big.NewInt(0)
+	wrapper := staking.ValidatorWrapper{}
+	wrapper.Validator = *v
+	wrapper.Delegations = []staking.Delegation{
+		staking.NewDelegation(v.Address, msg.Amount),
+	}
+	wrapper.Snapshot.Epoch = epoch
+	wrapper.Snapshot.NumBlocksSigned = zero
+	wrapper.Snapshot.NumBlocksToSign = zero
+	return wrapper.SanityCheck()
+}
+
+// VerifyEditValidatorStakingMsg verifies the edit validator message using
+// the stateDB and chainContext if they are provided.
+// It is not necessary to provide the stateDB and/or chainContext; validation that
+// does not use said components will still run.
+//
+// Note that these checks never update the stateDB, it only reads from stateDB.
+func VerifyEditValidatorStakingMsg(
+	stateDB vm.StateDB, chainContext ChainContext, msg *staking.EditValidator,
+) error {
+	if stateDB == nil {
+		return verifyEditValidatorStakingMsgNoState(msg)
+	}
+	if !stateDB.IsValidator(msg.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+	wrapper := stateDB.GetStakingInfo(msg.ValidatorAddress)
+	if wrapper == nil {
+		return errValidatorNotExist
+	}
+	if err := staking.UpdateValidatorFromEditMsg(&wrapper.Validator, msg); err != nil {
+		return err
+	}
+	newRate := wrapper.Validator.Rate
+	if newRate.GT(wrapper.Validator.MaxRate) {
+		return errCommissionRateChangeTooHigh
+	}
+	if chainContext != nil {
+		// TODO: make sure we are reading from the correct snapshot
+		snapshotValidator, err := chainContext.ReadValidatorSnapshot(wrapper.Address)
+		if err != nil {
+			return err
+		}
+		rateAtBeginningOfEpoch := snapshotValidator.Validator.Rate
+
+		if newRate.Sub(rateAtBeginningOfEpoch).Abs().GT(wrapper.Validator.MaxChangeRate) {
+			return errCommissionRateChangeTooFast
+		}
+	}
+	return wrapper.SanityCheck()
+}
+
+// VerifyDelegateStakingMsg verifies the delegate message using the stateDB if
+// it is provided. Otherwise, it will only check components that don't require it.
+//
+// Note that these checks never update the stateDB, it only reads from stateDB.
+func VerifyDelegateStakingMsg(stateDB vm.StateDB, msg *staking.Delegate) error {
+	if msg.Amount.Sign() == -1 {
+		return errNegativeAmount
+	}
+	if stateDB == nil {
+		return nil
```

Comment:
If there is no state, then should return error. Otherwise, it can be a false negative.

Reply:
okay will return an error